src/LinuxTracing/PerfEvent.h

```diff
@@ -142,6 +142,25 @@ struct UretprobesWithReturnValuePerfEventData {
 };
 using UretprobesWithReturnValuePerfEvent = TypedPerfEvent<UretprobesWithReturnValuePerfEventData>;
 
+// UserSpaceFunctionEntryPerfEvent and UserSpaceFunctionExitPerfEvent don't correspond to records
+// produced by perf_event_open, but rather to the FunctionEntry and FunctionExit protos emitted by
+// user space instrumentation. They are processed in LinuxTracing because, just like with uprobes,
+// unwinding and dynamic instrumentation need to work together.
+struct UserSpaceFunctionEntryPerfEventData {
```

Comment:
So to ensure that I understand it correctly:
We have the userspace events (protos) that we receive, convert into "PerfEvents" and then store those into the event buffer, such that userspace and callstack events are ordered when processed by the uprobes processor. Is that correct?

Reply:
Precisely. The "hacky" architecture way.